templates/angular/src/app/app.component.spec.ts

```diff
@@ -0,0 +1,107 @@
+import { TestBed, ComponentFixture, waitForAsync } from '@angular/core/testing'
+import { FormsModule } from '@angular/forms'
+import { By } from '@angular/platform-browser'
+
+import { setUpTestConnection, generateUniqueString, createAccount } from 'near-api-js/test/test-utils'
+import { WalletConnection } from 'near-api-js'
+
+import { AppComponent } from './app.component'
+import { WINDOW } from './services/window.service'
+
+describe('AppComponent', () => {
+  let app: AppComponent
+  let fixture: ComponentFixture<AppComponent>
+  let spyWindow: jasmine.SpyObj<Window>
+  let mockWindow
+
+  beforeAll(waitForAsync(async (): Promise<void> => {
+    const near = await setUpTestConnection()
+    const walletConnection = new WalletConnection(near, 'test')
+
+    mockWindow = {
+      accountId: 'test.near',
+      contract: {
+        account: createAccount(near),
+        contractId: generateUniqueString('test'),
+        getGreeting(): string {
+          return 'Hello'
+        },
+        setGreeting(): void {}
+      },
+      walletConnection
+    }
+  }))
+
+  beforeEach(waitForAsync(() => {
+    TestBed.configureTestingModule({
+      imports: [FormsModule],
+      declarations: [
+        AppComponent
+      ],
+      providers: [
+        {
+          provide: WINDOW,
+          useValue: mockWindow
+        }
+      ]
+    }).compileComponents()
+  }))
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(AppComponent)
+    app = fixture.componentInstance
+    spyWindow = TestBed.inject(WINDOW) as jasmine.SpyObj<Window>
+    fixture.detectChanges()
+  })
+
+  it('should create the app', () => {
+    expect(app).toBeTruthy()
+  })
+
+  describe('signedOut', () => {
+    it('should display welcome message', () => {
+      const compiled = fixture.nativeElement
+      const h1 = compiled.querySelector('main h1')
+
+      expect(h1.textContent).toContain('Welcome to NEAR!')
+    })
+
+    it('should call the `login` method on `Sign in` action', () => {
+      const btn = fixture.debugElement.query(By.css('button'))
```

Comment:
Copying my old message here:

> In the interest of setting up a project for success at it grows larger, could this button be found by using its innerHTML, rather than a CSS selector? I would imagine that an app could have lots of button.link elements as it grows

Selecting any `button` seems like it might run into problems even more quickly, as people add new buttons to their apps. Can you select by `innerHTML`?

Reply:
I fixed the selector for the `Sign in` button.